internal/api/handler.go

```diff
@@ -84,57 +103,141 @@ func (h *Handler) handleAuthorize(c *Context, w http.ResponseWriter, r *http.Req
+	redirectURI, err := url.Parse(chimeraAuthZState.RedirectURI)
 	if err != nil {
 		c.Logger.Error("Failed to parse destination site URL")
 		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
+	// Update redirect URI with params returned from OAuth provider such as `authorization_code`.
 	originalState := state[:len(state)-stateExtensionLength]
-
 	query := r.URL.Query()
 	query.Set("state", originalState)
-
 	redirectURI.RawQuery = query.Encode()
-	c.Logger.Info("Redirecting authorization callback")
 
-	http.Redirect(w, r, redirectURI.String(), http.StatusFound)
+	redirectURIStr := redirectURI.String()
+
+	verificationToken := util.NewRandomString(32)
+	err = h.stateCache.SetRedirectURI(state, cache.AuthorizationState{RedirectURI: redirectURIStr, AuthorizationVerificationToken: verificationToken})
+	if err != nil {
+		c.Logger.Error("Failed to update redirect URL in cache")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	c.Logger.Info("Redirecting authorization callback to confirmation")
+	http.Redirect(w, r, fmt.Sprintf("%s/v1/%s/%s/auth/chimera/confirm?state=%s", h.baseURL, c.OAuthApplication.Provider, c.OAuthApplication.Identifier, state), http.StatusFound)
 }
 
-func (h *Handler) handleTokenExchange(c *Context, w http.ResponseWriter, r *http.Request) {
-	app, found := h.getOAuthApp(c, r)
-	if !found {
-		w.WriteHeader(http.StatusNotFound)
+func (h *Handler) handleGetConfirmAuthorization(c *OAuthAppContext, w http.ResponseWriter, r *http.Request) {
+	c.Logger = loggerWithAppFields(c.Logger, c.OAuthApplication)
+	c.Logger.Info("Handling request for confirmation of Chimera authZ")
+
+	state := r.URL.Query().Get("state")
+	fmt.Println("Handle confirm state: ", state)
+
+	chimeraAuthZState, err := h.getAuthZStateFromCache(state)
+	if err != nil {
+		c.Logger.WithError(err).Error("Failed to get redirect URI from cache")
+		w.WriteHeader(statuserr.ErrToStatus(err))
 		return
 	}
-	c.Logger = c.Logger.WithField("provider", app.Provider).
-		WithField("app", app.Identifier)
+
+	// Strip URL query for cleaner display
+	strippedURL, err := stripURLQuery(chimeraAuthZState.RedirectURI)
+	if err != nil {
+		c.Logger.Error("Failed to strip query from redirect URL")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	data := AuthFormData{
+		RedirectURL:    strippedURL,
+		ConfirmAuthURL: fmt.Sprintf("%s/v1/%s/%s/auth/chimera/confirm?state=%s", h.baseURL, c.OAuthApplication.Provider, c.OAuthApplication.Identifier, state),
+		ProviderName:   c.OAuthApplication.Provider.DisplayName(),
+		ProviderURL:    c.OAuthApplication.Provider.HomepageURL(),
+		CancelAuthURL:  fmt.Sprintf("%s/v1/auth/chimera/cancel?state=%s", h.baseURL, state),
+		CsrfField:      csrf.TemplateField(r),
+	}
+
+	c.Logger.Info("Displaying authZ confirmation from")
+	http.SetCookie(w, h.chimeraAuthZTokenCookie(chimeraAuthZState.AuthorizationVerificationToken))
+	w.Header().Set("X-Frame-Options", "DENY")
```

Comment:
This should probably be applied to all responses in middleware. For defense in depth there should also be a CSP header, something like `Content-Security-Policy: default-src 'self'; font-src fonts.gstatic.com; style-src 'self' fonts.googleapis.com`.